File: 84.柱状图中最大的矩形.c
/*
 * @lc app=leetcode.cn id=84 lang=c
 *
 * [84] 柱状图中最大的矩形
 */

// @lc code=start

int largestRectangleArea(int *heights, int heightsSize)
{
  // 法一：双指针法
  // int i, left, right, max = 0, h, w;
  // for (i = 0; i < heightsSize; i++)
  // {
  //   left = i;
  //   right = i;
  //   for (; left >= 0; left--)
  //     if (heights[left] < heights[i])
  //       break;
  //   for (; right < heightsSize; right++)
  //     if (heights[right] < heights[i])
  //       break;
  //   w = right - left - 1;
  //   h = heights[i];
  //   max = max > w * h ? max : w * h;
  // }

  // return max;

  // 法二：单调栈
  int top, *stack, max = 0, i, mid, w, h, *newHeight;
  stack = (int *)malloc(sizeof(int) * (heightsSize + 2));
  newHeight = (int *)malloc(sizeof(int) * (heightsSize + 2));
  for (i = 1; i < heightsSize + 1; i++)
    newHeight[i] = heights[i - 1];
  newHeight[0] = 0;
  newHeight[heightsSize + 1] = 0;

  top = 0;
  stack[top++] = 0;
  for (i = 1; i < heightsSize + 2; i++)
  {
    while (top != 0 && newHeight[i] < newHeight[stack[top - 1]])
    {
      mid = stack[--top];
      w = i - stack[top - 1] - 1;
      h = newHeight[mid];
      max = max > w * h ? max : w * h;
    }
    stack[top++] = i;
  }

  return max;
}
// @lc code=end

File: store/84.柱状图中最大的矩形.js
/*
 * @lc app=leetcode.cn id=84 lang=javascript
 *
 * [84] 柱状图中最大的矩形
 */

// @lc code=start
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function (heights) {
  // 法一 超时
  // let minIndex = new Array(heights.length).fill(1),
  //   maxIndex = new Array(heights.length).fill(1),
  //   sum = 0;

  // minIndex[0] = -1;
  // maxIndex[heights.length - 1] = heights.length;
  // for (let i = heights.length - 1; i > 0; i--) {
  //   let t = i - 1;
  //   while (t >= 0 && heights[t] >= heights[i]) t--;
  //   minIndex[i] = t;
  // }

  // for (let i = 0; i < heights.length; i++) {
  //   let t = i + 1;
  //   while (t < heights.length && heights[t] >= heights[i]) t++;
  //   maxIndex[i] = t;
  // }

  // for (let i = 0; i < heights.length; i++) {
  //   const w = maxIndex[i] - minIndex[i] - 1;
  //   const h = heights[i];
  //   if (w > 0) sum = Math.max(sum, w * h);
  // }

  // return sum;

  // 法二
  // let minIndex = new Array(heights.length).fill(1),
  //   maxIndex = new Array(heights.length).fill(1),
  //   sum = 0;

  // minIndex[0] = -1;
  // maxIndex[heights.length - 1] = heights.length;
  // for (let i = 1; i < heights.length; i++) {
  //   let t = i - 1;
  //   while (t >= 0 && heights[t] >= heights[i]) t = minIndex[t];
  //   minIndex[i] = t;
  // }

  // for (let i = heights.length - 1; i >= 0; i--) {
  //   let t = i + 1;
  //   while (t < heights.length && heights[t] >= heights[i]) t = maxIndex[t];
  //   maxIndex[i] = t;
  // }

  // for (let i = 0; i < heights.length; i++) {
  //   const w = maxIndex[i] - minIndex[i] - 1;
  //   const h = heights[i];
  //   if (w > 0) sum = Math.max(sum, w * h);
  // }

  // return sum;

  // 法三
  let stack = [],
    top = 0,
    max = 0;
  stack[top++] = 0;
  heights.unshift(0);
  heights.push(0);
  for (let i = 1; i < heights.length; i++) {
    while (heights[i] < heights[stack[top - 1]]) {
      const item = stack[--top];
      const w = i - stack[top - 1] - 1;
      const h = heights[item];
      max = Math.max(max, w * h);
    }
    stack[top++] = i;
  }
  return max;
};
// @lc code=end
